File: Course_1-Algorithmic_Toolbox/Week-2/Excercise_Challenges/6_last_digit_of_the_sum_of_fibonacci_numbers/fibonacci_sum_last_digit.py
PISANO = 60

def fibonacci_sum_fast(n):
    if n < 2:
        return n

    n %= PISANO

    F = [1, 1]
    for _ in range(n):
        F.append((F[-1] + F[-2]) % 10)

    return (F[-1] - 1) % 10

def fibonacci_sum_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum      = 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum += current

    return sum % 10

if __name__ == '__main__':
    n = int(input())
    print(fibonacci_sum_fast(n))

File: Course_1-Algorithmic_Toolbox/Week-2/Excercise_Challenges/6_last_digit_of_the_sum_of_fibonacci_numbers/fibonacci_sum_last_digit.cpp
#include <iostream>
#define PISANO 60 // pisano period for 10
typedef long long ll;
using namespace std;

int fibonacci_sum_fast(ll n)
{
    if(n < 2)
        return (int)n;

    n %= PISANO;
    int prev = 1;
    int curr = 1;

    for(int i = 0; i < n; ++i){
        int temp = (prev + curr) % 10;
        prev = curr;
        curr = temp;
    }

    return (!curr? 9 : (curr - 1) % 10);
}

int fibonacci_sum_naive(long long n)
{
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;
    long long sum      = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
        sum += current;
    }

    return sum % 10;
}

int main(void)
{
    ll n;
    cin >> n;
    cout << fibonacci_sum_fast(n) << "\n";

    return 0;
}
